### PR TITLE
override watchWith in EffectfulActorContext, fixes #24924

### LINKED ISSUE #24924: Source.unfoldResource runs `close` twice.
Consider the following snippet:

```scala
import akka.actor.ActorSystem
import akka.stream.ActorMaterializer
import akka.stream.scaladsl.{Sink, Source}
import akka.stream.testkit.scaladsl.TestSink

object Test extends App {
  implicit val system = ActorSystem()
  implicit val materializer = ActorMaterializer()

  var count = 0

  def makeSource = {
    Source.unfoldResource[Int, Int](
      () => {
        count = count + 1
        println("obtaining: " + count)
        count
      },
      c => if(c > 5) {
        println("Limit reached!")
        throw new Exception("")
      } else {
        Some(c)
      },
      c => {
        println("releasing: " + c)
        count = count - 1
      }
    ).map(_ * 2)
     .recover { case e => 23 }
  }

  1.to(10).map(_ => makeSource).foreach { s =>
    val p = s.runWith(TestSink.probe)
    p.request(1) // Generate demand.
  }

}
```

Output:

```
obtaining: 1
obtaining: 2
obtaining: 3
obtaining: 4
obtaining: 5
obtaining: 6
Limit reached!
releasing: 6
releasing: 6
obtaining: 5
obtaining: 6
Limit reached!
releasing: 6
releasing: 6
obtaining: 5
obtaining: 6
Limit reached!
releasing: 6
releasing: 6
```

The code creates 10 streams, each of which tries to obtain a resource using `Source.unfoldResource`, does some processing (here a `map`) and `recovers` from any errors. The resource is limited, so after a certain point some of these streams fail to obtain the resource  resulting in `close` passed to `Source.unfoldResource` being called twice in every such case.

Is this the expected and correct behaviour? I would expect it to relese the resource only once and the correct output to be the following:

```
obtaining: 1
obtaining: 2
obtaining: 3
obtaining: 4
obtaining: 5
obtaining: 6
Limit reached!
releasing: 6
obtaining: 6
Limit reached!
releasing: 6
obtaining: 6
Limit reached!
releasing: 6
...
```

Also, moving the `if` further downstream as in the following snippet makes it work as expected:

```scala
// ...
  def makeSource = {
    Source.unfoldResource[Int, Int](
      () => {
        count = count + 1
        println("obtaining: " + count)
        count
      },
      Some(_),
      c => {
        println("releasing: " + c)
        count = count - 1
      }
    ).map { c =>
      if(c > 5) {
        println("Limit reached!")
        throw new Exception("")
      } else {
        c * 2
      }
    }
    .recover { case e => 23 }
  }
// ...
```

### PR DESCRIPTION
**Pull request by rkuhn**

### CODE REVIEW

**Comment by rkuhn** [comment]
fixes #24824

**Review by chbatey — APPROVED** [review]
LGTM

**Review by johanandren — APPROVED** [review]
LGTM with one question

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/EffectfulActorContext.scala [hunk 1]
```diff
@@ -50,6 +50,10 @@ import scala.concurrent.duration.{ Duration, FiniteDuration }
     effectQueue.offer(Watched(other))
     super.watch(other)
   }
+  override def watchWith[U](other: ActorRef[U], msg: T): Unit = {
+    effectQueue.offer(Watched(other))
```
[the remark this answers is not in the document]
Comment (rkuhn) [reply akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/EffectfulActorContext.scala]: `watchWith` is semantically a `watch` with an additional signal handler that injects the specified message, so I’d consider the distinction an internal detail of the actor that should not leak into the test. Of course, this assumes that we implement #24825 and that the mechanism will indeed inject the required message upon other actor termination.

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**